src/Database/LSMTree/Internal/RawOverflowPage.hs

```diff
@@ -0,0 +1,161 @@
+    , invariant page
+    = page
+
+    | otherwise
+    = makeRawOverflowPageCopy ba off len
+
+makeRawOverflowPageCopy
+    :: ByteArray  -- ^ bytearray
+    -> Int        -- ^ offset in bytes into the bytearray
+    -> Int        -- ^ length in bytes
+    -> RawOverflowPage
+makeRawOverflowPageCopy ba off len =
+    assert (len >= 0 && len < 4096 && len <= sizeofByteArray ba - off) $
+    (\page -> assert (invariant page) page) $
+    RawOverflowPage 0 $ runByteArray $ do
+      mba <- newPinnedByteArray 4096
+      let suffixlen = min 4096 len -- would only do anything if assertions off
+      copyByteArray mba 0 ba off suffixlen
+      when (suffixlen < 4096) $ fillByteArray mba suffixlen (4096-suffixlen) 0
+      return mba
+
+-- | Create a 'RawOverflowPage' without copying. The byte array and offset must
+-- satisfy the invariant for 'RawOverflowPage'.
+--
+unsafeMakeRawOverflowPage
+    :: ByteArray  -- ^ bytearray, must be pinned and contain 4096 bytes (after offset)
+    -> Int        -- ^ offset in bytes
+    -> RawOverflowPage
+unsafeMakeRawOverflowPage ba off =
+    assert (invariant page) page
+  where
+    page = RawOverflowPage off ba
+
+-- | Convert 'RawBytes' representing the \"overflow\" part of a value into one
+-- or more 'RawOverflowPage's.
+--
+-- This will avoid copying where possible.
+--
+rawBytesToOverflowPages :: RawBytes -> [RawOverflowPage]
+rawBytesToOverflowPages (RawBytes (P.Vector off len ba))
+  | isByteArrayPinned ba
+  = rawBytesToOverflowPagesPinned off len ba
+  | otherwise
+  = rawBytesToOverflowPagesUnpinned off len ba
+
+rawBytesToOverflowPagesPinned :: Int -> Int -> ByteArray -> [RawOverflowPage]
+rawBytesToOverflowPagesPinned !off !len !ba
+      | len >= 4096
+      , let !page = unsafeMakeRawOverflowPage ba off
+      = page : rawBytesToOverflowPagesPinned (off+4096) (len-4096) ba
+
+      | len <= 0
+      = []
+
+      | otherwise -- > 0 && < 4096
+                  -- have to copy the partial last page
+      , let !page = makeRawOverflowPageCopy ba off len
+      = page : []
+
+-- | Not pinned, in principle shouldn't happen much because if the value
+-- is big enough to overflow then it's big enough to be pinned.
+-- It is possible however if a page has a huge key and a small value.
+--
+-- Unfortunately, with GHC versions 9.6.x we also get this because the meaning
+-- of pinned has changed. Sigh.
+-- See <https://gitlab.haskell.org/ghc/ghc/-/issues/22255>
+--
+rawBytesToOverflowPagesUnpinned :: Int -> Int -> ByteArray -> [RawOverflowPage]
+rawBytesToOverflowPagesUnpinned !off !len !ba =
+    let !lenPages = roundUpToPageSize len
+        ba'       = runByteArray $ do
+                      mba <- newPinnedByteArray lenPages
+                      copyByteArray mba 0 ba off len
+                      fillByteArray mba len (lenPages-len) 0
+                      return mba
+     in assert (sizeofByteArray ba' == lenPages) $
+        assert (lenPages `mod` 4096 == 0) $
```

Comment:
This might fit better as an assertion inside `roundUpToPageSize`

Reply:
I've changed it to assert on the result rather than input. The real condition we want to check is that the conversion to overflow pages now works without copying. So we assert that.